src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs

```diff
@@ -65,12 +65,45 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
             Ensure.IsNotNull(connection, nameof(connection));
             Ensure.IsNotNull(description, nameof(description));
 
+            var firstStepResult = description.IsMasterResult.SpeculativeAuthenticate;
+
+            if (firstStepResult == null)
+            {
+                Authenticate(connection, description, previousStep: null, previousStepResult: null, cancellationToken);
+            }
+            else
+            {
+                (var firstStep, _) = CreateFirstStepAndCommand(_mechanism, connection, conversation: null, description);
```

Comment:
Named tuples: we could refactor this to use a small class or struct. I've created a thread in our channel to discuss the inclusion of the ValueTuple dependency due to the sensitive nature of this discussion (as explained in the thread)

Reply:
Resolved offline!